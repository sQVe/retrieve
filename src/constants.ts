import { Container } from './types'

export const defaultContainer: Container = {
  url: '',
  init: {
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    method: 'GET',
  },
  options: { resolveAs: 'JSON' },
}
