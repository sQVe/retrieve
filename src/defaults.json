{
  "init": {
    "headers": {
      "Accept": "application/json",
      "Content-Type": "application/json"
    },
    "method": "GET"
  },
  "opts": {
    "rawPayload": false,
    "responseAs": "json"
  }
}
