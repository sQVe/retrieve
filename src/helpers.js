import { compose } from './utility'

/**
 * createQuery :: {k:v} -> string
 */
export function createQuery(params) {
  if (!params) return ''

  const arr = Object.entries (params).map (([k, v]) =>
    [k, encodeURIComponent (v)].join ('=')
  )

  return arr.length ? arr.join ('&') : ''
}

/**
 * sanitizeUrl :: string -> string
 */
export function sanitizeUrl(url) {
  const addTrailingSlash = x => (x.endsWith ('/') ? x : `${x}/`)
  const removeLeadingSlash = x => (x.startsWith ('/') ? x.substring (1) : x)

  return url
    ? compose (
        addTrailingSlash,
        removeLeadingSlash
      ) (url)
    : ''
}
