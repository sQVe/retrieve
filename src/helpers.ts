import {
  Container,
  NotNil,
  OnResponse,
  QueryParams,
  ResolveAs,
  ResolveMethod,
} from './types'
import { isNonEmptyString } from './utility'

/** combineUrls :: [String] -> String */
export const combineUrls = (urls: (string | undefined)[]) => {
  const addTrailingSlash = (x: string) => (x.endsWith('/') ? x : `${x}/`)
  const removeLeadingSlash = (x: string) =>
    x.startsWith('/') ? x.substring(1) : x

  const validUrls = (urls || []).filter(isNonEmptyString)

  if (validUrls.length === 0) return ''
  if (validUrls.length === 1) return validUrls[0]

  return validUrls.reduce(
    (acc, url) => addTrailingSlash(acc) + removeLeadingSlash(url)
  )
}

/** createQuery :: {k: v} -> String */
export const createQuery = (params: QueryParams) => {
  const arr = Object.entries(params || {}).map(([k, v]) =>
    [k, encodeURIComponent(v)].join('=')
  )

  return arr.length ? arr.join('&') : ''
}

/** getResolveAsMethodName :: String -> String */
export const getResolveAsMethodName = (resolveAs?: ResolveAs) => {
  const methodNameLookup: { [index: string]: ResolveMethod } = {
    arraybuffer: 'arrayBuffer',
    blob: 'blob',
    formdata: 'formData',
    json: 'json',
    response: 'response',
    text: 'text',
  }

  if (resolveAs == null) return methodNameLookup.response

  const caseSafeResolveAs = resolveAs.toLowerCase()
  return methodNameLookup[caseSafeResolveAs] || resolveAs
}

/** handleResponse :: Response r => (r a -> void) -> r a -> void */
export const handleResponse = (onResponse?: OnResponse) => (res: Response) =>
  onResponse && onResponse(res)

/** preparePayload :: Container -> a -> b */
export const preparePayload = (container: Container) => (payload: NotNil) => {
  const { payloadAs } = container.options

  return payload != null && payloadAs != null ? payloadAs(payload) : payload
}

/** resolveResponse :: Response a -> String -> Promise b */
export const resolveResponse = (res: Response) => async (
  resolveAs?: ResolveAs
) => {
  const resolveAsMethodName = getResolveAsMethodName(resolveAs)

  if (resolveAsMethodName === 'response') return res
  if (resolveAsMethodName === 'json') {
    const text = await res.clone().text()

    if (text.length === 0) return undefined // eslint-disable-line fp/no-nil
    return res.json()
  }

  return res[resolveAsMethodName]()
}

/** combineContainers :: Container -> Container -> Container */
export const combineContainers = (a: Partial<Container>) => (
  b: Partial<Container>
): Container => ({
  url: combineUrls([a.url, b.url]),
  init: { ...(a.init || {}), ...(b.init || {}) },
  options: { ...(a.options || {}), ...(b.options || {}) },
})
