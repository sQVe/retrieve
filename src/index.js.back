import { createQuery, urlGuard } from './utility'

const defaults = {
  init: {
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    method: 'GET',
  },
  opts: {
    bindContext: true,
    rawPayload: false,
    responseAs: 'json',
  },
}

export default class Righttp {
  constructor(url, init, opts) {
    this.url = urlGuard (url)
    this.init = { ...defaults.init, ...(init || {}) }
    this.opts = { ...defaults.opts, ...(opts || {}) }

    if (this.opts.bindContext) {
      ;[/* 'delete', 'get', 'post', 'put', */ 'request'].forEach (f => {
        this[f] = this[f].bind (this)
      })
    }
  }

  applyDefaults(url, init) {
    return [this.url + urlGuard (url), { ...this.init, ...(init || {}) }]
  }

  constructInit(key, val, acc) {
    let x = val

    if (acc === undefined) x = key ? this.init[key] : this.init
    if (typeof x === 'function') x = x ()
    if (typeof x !== 'object') return x

    return Object.entries (x).reduce (
      (a, [k, v]) => ({ ...a, [k]: this.constructInit (null, v, a) }),
      acc || {}
    )
  }

  currentContentType() {
    return this.constructInit ('headers')['Content-Type']
  }

  forwardToStatusListener(res) {
    const onStatus = this.opts.onStatus || {}
    const { status } = res

    const isInStatusRange = key => {
      const [x, y] = key.split ('-')

      return x && y && +x < status && +y > status
    }

    Object.keys (onStatus).forEach (
      k => (onStatus[status] || isInStatusRange (k)) && onStatus[k] (res)
    )
  }

  async parseResponse(res) {
    const responseAs = (this.opts && this.opts.responseAs) || 'response'

    switch (responseAs) {
      case 'json':
        try {
          const text = await res.text ()
          if (text.length > 0) return JSON.parse (text)
          return undefined
        } catch (err) {
          throw err
        }
      case 'response':
        return res
      default:
        return res[responseAs] ()
    }
  }

  preparePayload(data) {
    if (this.opts.rawPayload) return data

    switch (this.currentContentType ()) {
      case 'application/json':
        return JSON.stringify (data)
      case 'application/x-www-form-urlencoded':
        return createQuery (data)
      case 'multipart/form-data':
        return Object.entries (data).reduce ((acc, entry) => {
          acc.append (...entry)
          return acc
        }, new FormData ())
      default:
        return data
    }
  }

  async request(url, init) {
    const [preparedUrl, preparedInit] = this.applyDefaults (url, init)

    if (preparedUrl.length === 0)
      throw new Error ('Righttp needs an URL to make an request')

    const res = await fetch (preparedUrl, preparedInit)

    this.forwardToStatusListener (res)
    if (res.ok) return this.parseResponse (res)
    throw res
  }
}
