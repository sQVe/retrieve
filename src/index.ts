export { righttp, righttp as default } from './righttp'
