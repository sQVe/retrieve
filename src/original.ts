/* eslint-disable */

type Params = {
  [key: string]: string
}

type StatusCallbacks = {
  [key: string]: (res: Response) => void
}

interface Init extends RequestInit {
  method: Method
  onStatus?: StatusCallbacks
  responseAs: ResponseAs
}

export enum Method {
  connect = 'CONNECT',
  delete = 'DELETE',
  get = 'GET',
  head = 'HEAD',
  options = 'OPTIONS',
  patch = 'PATCH',
  post = 'POST',
  put = 'PUT',
  trace = 'TRACE',
}

export enum ResponseAs {
  arrayBuffer = 'arrayBuffer',
  blob = 'blob',
  formData = 'formData',
  json = 'json',
  response = 'response',
  text = 'text',
}

export class Http {
  private _url: string
  private _init: Init

  constructor(url = '', init: Partial<Init> = {}) {
    this._url = this._urlGuard(url)
    this._init = {
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      method: Method.get,
      responseAs: ResponseAs.json,
      ...init,
    }
  }

  public get url() {
    return this._url
  }

  public get init() {
    return this._init
  }

  public delete(url = '', data?: object): Promise<any> {
    return this.request(url, {
      method: Method.delete,
      body: this._preparePayload(data),
    })
  }

  public get(url = '', params?: Params): Promise<any> {
    return this.request(
      this._urlGuard(url, true) +
        (params == null ? '' : this._combineToQuery(params))
    )
  }

  public patch(url = '', data: object): Promise<any> {
    return this.request(url, {
      method: Method.patch,
      body: this._preparePayload(data),
    })
  }

  public post(url = '', data: object): Promise<any> {
    return this.request(url, {
      method: Method.post,
      body: this._preparePayload(data),
    })
  }

  public put(url = '', data: object): Promise<any> {
    return this.request(url, {
      method: Method.put,
      body: this._preparePayload(data),
    })
  }

  public async request(url: string = '', init?: Partial<Init>): Promise<any> {
    const [reqUrl, reqInit] = this._applyDefaults(url, init)

    if (reqUrl.length === 0)
      throw new Error('Http needs an URL to make an request')

    const res = await fetch(reqUrl, reqInit)

    this._handleStatus(res)
    if (res.ok) {
      return reqInit.responseAs === 'response'
        ? res
        : this._parseResponse(res, reqInit.responseAs)
    }
    throw res
  }

  public fork(fragment?: string, init?: Partial<Init>): Http {
    return new Http(this._url + this._urlGuard(fragment, true), init)
  }

  private _applyDefaults(url = '', init: Partial<Init> = {}): [string, Init] {
    return [
      this.url + this._urlGuard(url, true),
      { ...this.init, ...init } as Init,
    ]
  }

  private _combineToQuery(params: Params): string {
    const arr = Object.keys(params).map(
      k => k + '=' + encodeURIComponent(params[k])
    )

    return '?' + arr.join('&')
  }

  private get _currentContentType(): string | undefined {
    return this._init.headers && this._init.headers['Content-Type']
  }

  private _handleStatus(res: Response): boolean {
    const onStatus = this._init.onStatus || {}

    const isInStatusRange = (key: string): boolean => {
      const range = key.split('-')

      return (
        range.length > 1 && +range[0] < res.status && +range[1] > res.status
      )
    }

    return Object.keys(onStatus)
      .map(key => {
        if (onStatus[res.status] || isInStatusRange(key)) {
          onStatus[key](res)
          return true
        }
        return false
      })
      .some(Boolean)
  }

  private async _parseResponse(
    res: Response,
    parseMethod: ResponseAs
  ): Promise<any> {
    switch (parseMethod) {
      // TODO: HtmlDocument?
      case 'json':
        const text = await res.clone().text()

        if (text.length > 0) return res.json()
        return {}
      default:
        return res[parseMethod]()
    }
  }

  private _urlGuard(url = '', isFragment = false): string {
    const addTrailingSlash = (x: string) => (x.endsWith('/') ? x : x + '/')
    const removeLeadingSlash = (x: string) =>
      x.startsWith('/') ? x.substring(1) : x

    if (url.length === 0) return ''
    if (isFragment) url = removeLeadingSlash(url)
    return addTrailingSlash(url)
  }

  private _preparePayload(data): any {
    switch (this._currentContentType) {
      case 'application/json':
        return JSON.stringify(data)
      case 'application/x-www-form-urlencoded':
        return this._combineToQuery(data).substr(1)
      default:
        return data
    }
  }
}
