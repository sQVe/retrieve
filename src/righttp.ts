import { Container, Init, Options, NotNil, QueryParams } from './types'
import {
  combineContainers,
  createQuery,
  handleResponse,
  loadPayload,
  resolveResponse,
} from './helpers'
import { defaultContainer } from './constants'

/** request :: Container -> Promise */
const request = async (container: Container) => {
  const {
    url,
    init,
    options: { onResponse, resolveAs },
  } = container

  if (url.length === 0)
    throw new Error('Righttp needs an URL to make a request.')

  const res = await fetch(url, init)

  handleResponse(onResponse)(res)

  if (!res.ok) throw new Error(`${res.status}: ${res.statusText}`)
  return resolveResponse(res)(resolveAs)
}

/**
 * Create a fetch wrapper with easy-to-use request methods.
 *
 * @param {string} url - The URL to preset.
 * @param {object} [init] - The fetch init to preset.
 * @param {object} [options] - The request options to preset.
 * @return {object} A righttp object with preset url, init and options.
 */
export function righttp(url: string, init?: Init, options?: Options) {
  const presetContainer = combineContainers(defaultContainer)({
    url,
    init,
    options,
  })
  const presetCombine = combineContainers(presetContainer)

  return {
    /**
     * Send a DELETE request.
     *
     * @function del
     * @param {string} url - The URL to request.
     * @param {*} [payload] - The request payload to attach.
     * @return {Promise<*>} Promise with request result.
     */
    del: (url: string, payload?: NotNil) => {
      const container = loadPayload(
        presetCombine({ url, init: { method: 'DELETE' } })
      )(payload)

      return request(container)
    },

    /**
     * Send a GET request.
     *
     * @function get
     * @param {string} url - The URL to request.
     * @param {object} [queryParams] - The query parameters to append to the URL.
     * @return {Promise<*>} Promise with request result.
     */
    get: (url: string, queryParams?: QueryParams) => {
      const container = presetCombine({
        url: queryParams == null ? url : url + '?' + createQuery(queryParams),
        init: { method: 'GET' },
      })

      return request(container)
    },

    /**
     * Send a PATCH request.
     *
     * @function patch
     * @param {string} url - The URL to request.
     * @param {*} payload - The request payload to attach.
     * @return {Promise<*>} Promise with request result.
     */
    patch: (url: string, payload: NotNil) => {
      const container = loadPayload(
        presetCombine({ url, init: { method: 'PATCH' } })
      )(payload)

      return request(container)
    },

    /**
     * Send a POST request.
     *
     * @function patch
     * @param {string} url - The URL to request.
     * @param {*} payload - The request payload to attach.
     * @return {Promise<*>} Promise with request result.
     */
    post: (url: string, payload: NotNil) => {
      const container = loadPayload(
        presetCombine({ url, init: { method: 'POST' } })
      )(payload)

      return request(container)
    },

    /**
     * The container which holds all preset url, init and option settings.
     *
     * @constant
     */
    preset: presetContainer,

    /**
     * Send a PUT request.
     *
     * @function patch
     * @param {string} url - The URL to request.
     * @param {*} payload - The request payload to attach.
     * @return {Promise<*>} Promise with request result.
     */
    put: (url: string, payload: NotNil) => {
      const container = loadPayload(
        presetCombine({ url, init: { method: 'PUT' } })
      )(payload)

      return request(container)
    },

    /**
     * Send a request.
     *
     * @function request
     * @param {string} url - The URL to request.
     * @param {object} [init] - The fetch init.
     * @param {object} [options] - The request options.
     * @return {Promise<*>} Promise with request result.
     */
    request: (url: string, init?: Init, options?: Options) =>
      request(presetCombine({ url, init, options })),
  }
}
