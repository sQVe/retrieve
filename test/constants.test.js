import { defaultContainer } from '../src/constants'

describe('constants', () => {
  beforeEach(() => {
    global.JSON = {
      stringify: jest.fn(),
    }
  })

  afterEach(jest.clearAllMocks)

  it('should have the correct default settings', () => {
    const payload = 'JSON'
    defaultContainer.options.payloadAs(payload)

    expect(defaultContainer).toMatchSnapshot()
    expect(JSON.stringify).toHaveBeenCalledTimes(1)
    expect(JSON.stringify).toHaveBeenCalledWith(payload)
  })
})
