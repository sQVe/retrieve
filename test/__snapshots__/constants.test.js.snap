// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constants should have the correct default settings 1`] = `
Object {
  "init": Object {
    "headers": Object {
      "Accept": "application/json",
      "Content-Type": "application/json",
    },
    "method": "GET",
  },
  "options": Object {
    "payloadAs": [Function],
    "resolveAs": "JSON",
  },
  "url": "",
}
`;
